### PR TITLE
Fix Combined Dataloader & unlabeled warmup loss to work correctly in Semi-SL case

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

Fixes some edge cases showing unintended behavior.
1. If the number of iters in the unlabeled dataloader is less than the number of iters in the labeled dataloader, find that combined dataloader are acting on the basis of the unlabeled dataloader and fix this.
2. Found that unlabeled warmup loss in Semi-SL was not working properly and had a shorter warmup time than expected, so fix this.

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

Model | Method | Test/Accuracy | E2E Time | Epochs
-- | -- | -- | -- | --
EfficientNet-B0 | SL     (Pseudo-Labeled dataset) | 0.5863 | 0:00:27 | 24
  | Semi-SL     (batch size: 16/48) | 0.4675 | 0:08:55 | 60
  | Semi-SL #Hot-Fix     (batch size: 16/48) | 0.6058 | 0:17:24 | 115

</div>

</body>

</html>

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by sungchul1 — APPROVED** [review]
LGTM, but could you explain if it is okay because Semi-SL takes a much longer training time than SL (about 39x based on the table) but has less performance improvement?

**Comment by harimkang** [comment]
> LGTM, but could you explain if it is okay because Semi-SL takes a much longer training time than SL (about 39x based on the table) but has less performance improvement?

For the table I used here, the experimental dataset is a bit unusual and relatively large, so I don't expect the change in PR to make much of a difference. Other than the change in the coefficient of the unlabeled loss, the schedule setting itself doesn't make a difference, so I expect it to behave more epochs if it can get more gain, but not less. I also expect this to show similar behavior to the minimum SL.

**Review by eunwoosh — APPROVED** [review]

**Comment by sungchul1** [comment]
> > LGTM, but could you explain if it is okay because Semi-SL takes a much longer training time than SL (about 39x based on the table) but has less performance improvement?
> 
> For the table I used here, the experimental dataset is a bit unusual and relatively large, so I don't expect the change in PR to make much of a difference. Other than the change in the coefficient of the unlabeled loss, the schedule setting itself doesn't make a difference, so I expect it to behave more epochs if it can get more gain, but not less. I also expect this to show similar behavior to the minimum SL.

Could you explain why you used unusual and relatively large dataset? To check the feasibility, I think common dataset should be used to accurately check improvement. And I'd like to suggest the experiments with two versions of datasets: #labeled > #unlabeled and #labeled < #unlabeled to see robust experimental results. What do you think?

**Comment by harimkang** [comment]
> Could you explain why you used unusual and relatively large dataset? To check the feasibility, I think common dataset should be used to accurately check improvement. And I'd like to suggest the experiments with two versions of datasets: #labeled > #unlabeled and #labeled < #unlabeled to see robust experimental results. What do you think?

This PR is a fix for the edge case, and there is little change in the typical Semi-SL case. This is an issue I discovered while experimenting with edge cases. Also, the results shared in the existing Semi-SL are mostly #labeled < #unlabeled, while the tables currently shared in this PR are #labeled > #unlabeled.
Anyway, I will confirm this with a benchmark test in the future. Thanks for the comment.
